messaging/src/main/java/io/lantern/messaging/tassis/Client.kt

```diff
@@ -0,0 +1,474 @@
+        }
+    }
+}
+
+/**
+ * A delegate for receiving lifecycle callbacks from a Client connection.
+ */
+interface ClientDelegate {
+    /**
+     * Called if there's an error connecting the Client
+     */
+    fun onConnectError(err: Throwable)
+
+    /**
+     * Called whenever we get a new Configuration from the server
+     */
+    fun onConfigUpdate(cfg: Messages.Configuration)
+
+    /**
+     * Called when a client closes. If the client closed due to an exception, err will be populated
+     * with the relevant exception. If the client closed while we were still in the process of
+     * connecting, the delegate will only receive a call to onConnectError.
+     */
+    fun onClose(err: Throwable? = null)
+}
+
+/**
+ * A delegate for receiving asynchronous callbacks on an authenticated client connection.
+ */
+interface AuthenticatedClientDelegate : ClientDelegate {
+    /**
+     * Called once the client has been connected
+     */
+    fun onConnected(client: AuthenticatedClient)
+
+    /**
+     * Used to sign the response to an AuthChallenge
+     */
+    fun signLogin(loginBytes: ByteArray): ByteArray
+
+    /**
+     * Called when tassis notifies us that one-time pre keys are getting low.
+     */
+    fun onPreKeysLow(numPreKeysRequested: Int)
+
+    /**
+     * Called when we receive an inbound message from another user device via tassis.
+     */
+    fun onInboundMessage(msg: InboundMessage)
+}
+
+interface AnonymousClientDelegate : ClientDelegate {
+    /**
+     * Called once the client has been connected
+     */
+    fun onConnected(client: AnonymousClient)
+}
+
+/**
+ * Client provides a mechanism for talking to a tassis service.
+ */
+abstract class Client<D : ClientDelegate>(
+    protected val delegate: D,
+    private val roundTripTimeoutMillis: Long
+) : MessageHandler {
+    private var transport: Transport? = null
+    private val msgSequence = AtomicInteger(1)
+    protected val pending = ConcurrentHashMap<Int, Callback<Any?>>()
+    protected val isConnecting = AtomicBoolean(true)
+    private val closed = AtomicBoolean()
+    private val timeoutChecker = Executors.newSingleThreadScheduledExecutor {
+        Thread(it, "client-timeout-checker")
+    }
+
+    override fun setTransport(transport: Transport) {
+        this.transport = transport
+    }
+
+    internal fun send(msg: Messages.Message, callback: Callback<*>? = null) {
+        val msgSequence = msg.sequence
+        if (callback != null) {
+            pending[msgSequence] = callback as Callback<Any?>
+        }
+        transport?.send(msg.toByteArray())
+        if (callback != null) {
+            timeoutChecker.schedule(
+                {
+                    pending.remove(msgSequence)?.let {
+                        // if any request times out, consider the whole connection bad and just close it
```

Comment:
Because if we get to the point of a timeout, the network connection is likely bad (e.g. was closed remotely without us knowing it) so it's better to reconnect to make sure we have a working connection.